### PR TITLE
[sui-execution] Introduce Executor interface

### PR DESCRIPTION
**Pull request by amnn**
## Description

This will be the uniform interface through which requests are made to the MoveVM within nodes.  It allows us to easily bundle multiple versions of the VM, adapter, natives and verifier into node software, and switch between them according to protocol config.

### Dynamic Interfaces

This change requires making previously static generics into trait objects, or shifting where the dynamism/ownership lies.

- Access to execution will be via the new `Executor` trait object.
- `TemporaryStore`'s access to the backing store (now via the `BackingStore` grouping trait).
- `ObjectRuntime` previously took a `Box<dyn ChildObjectResolver>`, now it's just a `&dyn ChildObjectResolver`.
- Execution engine's access to the store (now via the `ExecutionState` grouping trait).
- Type layout resolver's access to modules/packages etc (now via the `TypeLayoutStore` grouping trait).
- `LinkageView`'s access to the sui resolver traits is now via a trait object.

Some of these traits group multiple other traits, so extra functions have been added to do explicit conversion (e.g. from the vtable for an `A + B` trait object to the vtable for an `A` or a `B` trait object).

### Other Changes

- `InMemoryStorage::new` returns `Arc`
- Don't save native functions in `ExecutionComponents` -- create fresh, each time (hides native function related types in execution layer)
- Re-order function parameters for execution functions (only in newly introduced execution layer, will push this change down into execution implementations in a follow-up PR) to group together similar parameters.
-  genesis-related crates fully ported over to execution crate.
- `sui-replay` fully ported over to execution crate.
- De-nested implementation of `get_object_read` and `get_past_object_read` on `AuthorityState`.
- `sui-move/src/unit_test` creates a lazy `TEST_STORE` instead of relying on the `ObjectRuntime` to keep the store alive.
- `sui-replay`'s `LocalExec` implementation of all the various backing store traits return `SuiError` instead of `LocalExecError` (now that we use dynamic dispatch, the error type needs to be standardised).
- `get_package_objects` and `get_packages` helper functions on `BackingPackageStore` were previously default implemented trait functions, but they have turned into free functions.  This is because they are lifetime generic, and this interfered with the ability to use this trait (or any trait derived from it) as the basis of a trait object.
- Removed `temporary_store::empty_for_testing` and `::with_input_objects_for_testing` as unused, rather than port them (which would have been complicated and messy).
- Replaced `execute_transaction_to_effects` with `execute_transaction_to_effects_impl`, to avoid doubling up the number of functions in the new `Executor` interface.

## Test Plan

```
sui$ cargo simtest
sui$ env SUI_SKIP_SIMTESTS=1 cargo nextest run
```

Also manually test replay tool on a recent transaction digest:

```
sui/crates/sui-tool$ cargo run replay --rpc https://fullnode.testnet.sui.io:443 tx -t <digest>
```

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -1254,32 +1253,37 @@ impl AuthorityState {
             epoch_store.protocol_config(),
         );
         let (kind, signer, _) = transaction.execution_parts();
+
+        let silent = true;
         // don't bother with paranoid checks in dry run
         let enable_move_vm_paranoid_checks = false;
```
Comment (amnn) [inline crates/sui-core/src/authority.rs:1259]: Noticed a discrepency here, between dev inspect and dry run -- the former runs with paranoid checks enabled if the node is configured that way, whereas the latter does not.  Is that intentional? cc @tnowacki

File: crates/sui-core/src/authority/authority_store.rs [hunk 2]
```diff
@@ -1634,11 +1635,11 @@ impl AuthorityStore {
                 (init.0 + result.0, init.1 + result.1)
             })
         });
-        let mut layout_resolver = TypeLayoutResolver::new(move_vm, self.as_ref());
+        let mut layout_resolver = executor.type_layout_resolver(Box::new(self.as_ref()));
```
Comment (amnn) [inline crates/sui-core/src/authority/authority_store.rs:1638]: Noticed something here that I didn't fully understand.  Here, we are using the `AuthorityStore` directly as the store that `TypeLayoutResolver` queries for packages when resolving type layouts.  Above, we use the `package_cache`.  Why do we not use the package cache here? cc @lxfind .

File: sui-execution/src/executor.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    execution_mode::ExecutionResult,
+    gas::SuiGasStatus,
+    metrics::LimitsMetrics,
+    temporary_store::{InnerTemporaryStore, TemporaryStore},
+    transaction::{ProgrammableTransaction, TransactionKind},
+    type_resolver::LayoutResolver,
+};
+
+/// Abstracts over access to the VM across versions of the execution layer.
+pub trait Executor {
```
Comment (amnn) [inline sui-execution/src/executor.rs:26]: Originally, I intended to expose the `ExecutionMode` trait on this API boundary, but that's not possible because the `Executor` must be passed as a trait object, so the trait can't have generic type arguments (of which the `Mode` would have been one).  Instead the modes manifest as 4 different functions:

- `execute_transaction_to_effects` (`Normal`).
- `execute_system_transaction_to_effects` (`System`) -- potentially unused.
- `dev_inspect_transaction` (`DevInspect`).
- `update_genesis_state` (`Genesis`) -- runs a single PTB, rather than a `TransactionKind`.

**Review by amnn — COMMENTED** [review]
Notes for reviewers.

File: crates/sui-core/src/authority/authority_store.rs (hunk 2 above)
Comment (lxfind) [reply crates/sui-core/src/authority/authority_store.rs:1638]: That would be a question for @dariorussi. I think he added the layout resolver here.

File: sui-execution/src/executor.rs (hunk 3 above)
Comment (lxfind) [reply sui-execution/src/executor.rs:26]: Do you know why ExecutionMode was defined as a trait instead of just an Enum?

Comment (amnn) [reply sui-execution/src/executor.rs:26]: `ExecutionMode` comes with associated types, so it wouldn't be possible to encode it as an enum.  This is because when dev-inspecting a transaction, the return type of `execute_transaction_to_effects` changes.

File: crates/sui-core/src/authority/authority_store.rs (hunk 2 above)
Comment (dariorussi) [reply crates/sui-core/src/authority/authority_store.rs:1642]: not sure what to say here. I changed the code as it was, in other words it was using `self` before instead of the package cache and I left it that way. I wanted to raise the issue in the PR but then that PR got crazy and totally forgot.
Looking at the code it feels we should be able to use the cache which is really a wrapper around `self`.
Sorry I have no more to add here

File: crates/sui-core/src/authority.rs (hunk 1 above)
Comment (dariorussi) [reply crates/sui-core/src/authority.rs:1259]: I am surprised by this and I think we should remove it.
If that was for dev inspect I could relate more but for dry run I would expect all checks to be identical to those of a regular run.
I am curious to know if there is a reason for it

File: crates/sui-core/src/authority/authority_store.rs (hunk 2 above)
Comment (lxfind) [reply crates/sui-core/src/authority/authority_store.rs:1642]: https://github.com/MystenLabs/sui/pull/11529/files#diff-b07ea487376f67264505e88e54693c63f78e1720d0f9c928119d032a45925223 introduced this line, so it wasn't there before.

Comment (dariorussi) [reply crates/sui-core/src/authority/authority_store.rs:1642]: yes I know, however I was pointing out that the code changed as follows
```
-            total_sui += object.get_total_sui(self).unwrap() - object.storage_rebate;
+            total_sui +=
+                object.get_total_sui(&mut layout_resolver).unwrap() - object.storage_rebate;
+        }
```
in other words the call to `get_total_sui` was there with `self` and not `package_cache` and that is how I left it (maybe by mistake and should have "fixed" it).
So the question asked stands and it's whether that should take the package cache or not.
I think it should and wonder what is your take

File: sui-execution/src/executor.rs (hunk 3 above)
Comment (dariorussi) [reply sui-execution/src/executor.rs:26]: got to say that I always wondered about `Mode` and whether code refactoring could get rid of it. And this feels already a partial refactor, if nothing else exposing different functions for the specific case up the call stack.
I am obviously not asking to change anything but I wonder if you have an opinion about it

**Review by dariorussi — COMMENTED** [review]
just a comment and I'll come back to this to accept later, though it would be good for @tnowacki to have a look at this as well. And @lxfind too, though he has already seen this.
The overall refactor of traits, though still hard to follow for me, looks good, and everything looks fine.
I am not clear how this deals with multiple versions and I guess that is still to come.
The `AuthorityPerEpochStore` still has got 1 `ExecutionComponents` which carries an `Executor` which at the moment wraps a `MoveVM`. 
So as much as the trait allows for multiple implementation there is really no way yet to plug in multiple ones or multiple VMs, or is it?
This PR also made me realize the difficulty behind this issue (and maybe it is a difficulty for me only) which is that the VM and the Executor have totally different lifetimes.
If on one hand that makes me happy because, as you know, I think multiple version of a VM for the same "executor" are realistic and we should do them (I know later), on the other hand versioning a VM together (or as a side effect) of the "executor" (adapter) seems hard.
Don't know if I am making any sense and really curious to hear your view or that of others

**Comment by dariorussi** [comment]
> I am not clear how this deals with multiple versions and I guess that is still to come.

answering myself, maybe here you mean over different epoch (and it must be that way) in which case the single VM per epoch store makes sense

File: crates/sui-core/src/authority.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-core/src/authority.rs:1259]: Agreed -- I will do it in a follow-up.

File: crates/sui-core/src/authority/authority_store.rs (hunk 2 above)
Comment (amnn) [reply crates/sui-core/src/authority/authority_store.rs:1642]: Based on the resolution here (TBD), I'll also deal with this as a separate follow-up PR.  It does seem like we could use the package cache here though.

File: sui-execution/src/executor.rs (hunk 3 above)
Comment (amnn) [reply sui-execution/src/executor.rs:26]: The main thing that makes this difficult is dev inspect transactions which behave significantly differently (do extra work tracking intermediate results and changes to inputs, and return a different kind of execution result).  This makes sense because that's what `ExecutionMode` was originally created for.  My take is that the complexity here is inherent, and it's not worth trying to replace`ExecutionMode`, you'll just get something else equally complex.

**Comment by amnn** [comment]
> I am not clear how this deals with multiple versions and I guess that is still to come.

Yes that's right, it's still to come -- the way it will work is throuh the `executor` function in `sui-execution/lib.rs` -- this function will eventually read an execution version from `ProtocolConfig`, and based on that, it will pick the implementation of `dyn Executor` that it returns.

> This PR also made me realize the difficulty behind this issue (and maybe it is a difficulty for me only) which is that the VM and the Executor have totally different lifetimes.

I'm not sure I follow on this one -- `Executor` and `VM` are intended to have the same lifetimes, for the reason that you mention in your follow-up:

> answering myself, maybe here you mean over different epoch (and it must be that way) in which case the single VM per epoch store makes sense

And for this reason, I kind of hold the opposite view to below (i.e. that it's easier to version the VM with the executor together, rather than separately).

> on one hand that makes me happy because, as you know, I think multiple version of a VM for the same "executor" are realistic and we should do them (I know later), on the other hand versioning a VM together (or as a side effect) of the "executor" (adapter) seems hard.

With the work to remove `Session` etc, the line between the VM and the rest of the adapter layer is going to blur even further to the point where it will become less useful to version them separately because the adapter is so small.

**Review by dariorussi — APPROVED** [review]
@tnowacki take a look at this when you have a chance even if committed 
thanks
